06_IPC/03_Message_Passing.md

```diff
@@ -0,0 +1,176 @@
+
+```c
+ipc_endpoint* target = first_endpoint;
+//search the list for the endpoint we want
+acquire(&endpoints_lock);
+while (target != NULL) {
+    if (strcmp(target->name, ep_name) == 0)
+        break;
+    target = target->next;
+}
+
+release(&endpoints_lock);
+if (target == NULL)
+    return;
+```
+
+You may want to return an error here if the endpoint couldn't be found, however in our case we're simply discarding the message. 
+
+Now we'll need to allocate a buffer to store a copy of the message in, and copy the original message into this buffer.
+
+```c
+void* msg_copy = malloc(length);
+memcpy(msg_copy, buffer, length);
+```
+
+Why do we make a copy of the original message? Well if we dont, the sending process has to keep the original message around until the receiver has processed it. We don't have a way for the receiver to communicate that it's finished reading the message. By making a copy, we can return from `ipc_send()` as soon as the message is sent, regardless of when the message is read. Now the sending process is free to do what it wants with the memory holding the original message as soon as `ipc_send()` has completed.
+
+If you're performing this IPC as part of a system call from userspace, the memory containing the original message is unlikely to be mapped in the receiver's address space anyway, so we have to copy it into the kernel's address space, which is mapped in both processes.
+
+All that's left is to tell the receiver it has a message is available by placing the buffer address on the endpoint. Again, notice the use of the lock to prevent race conditions while we mess with the internals of the endpoint.
+
+```c
+acquire(&target->lock);
+target->msg_buffer = msg_copy;
+target->msg_length = length;
+release(&target->lock);
+```
+
+After the lock on the endpoint is released, the sent has been sent! Now it's up to the receiving thread to check the endpoint and set the buffer to NULL again.
+
+### Multiple Messages
+
+In theory this works, but we've overlooked one huge issue: what if there's already a message at the endpoint? You should handle this, and there's a couple of ways to go about it:
+
+- Allow for multiple messages to be stored on an endpoint.
+- Fail to send the message, instead returning an error code from `ipc_send()`.
+
+The first option is recommended, as it's likely there will be some processes that handle a lot of messages. Implementing this is left as an exercise to the user, but a simple implemenation might use a struct to hold each message (the buffer address and length) and a next field. Yes, more linked lists!
+
+Sending messages would now mean appending to the list instead of writing the buffer address as before. 
+
+## Receiving
+
+We can send messages, next let's look at receiving them. We're going to use a basic (and inefficient) example, but it shows how it could be done. 
```

Comment:
shouldn't be better adding now? Like:

"We have seen how to send messages, now let's look at how to receive them" (or something similar)

The current sentence give a sense of rush...

Reply:
yeah thats fair, I think I was rushing to get this done before getting off the train haha. I went with your version.